synapse/metrics/background_process_metrics.py

```diff
@@ -256,24 +256,55 @@ async def run() -> Optional[R]:
         return defer.ensureDeferred(run())
 
 
-F = TypeVar("F", bound=Callable[..., Awaitable[Optional[Any]]])
+P = ParamSpec("P")
 
 
-def wrap_as_background_process(desc: str) -> Callable[[F], F]:
-    """Decorator that wraps a function that gets called as a background
-    process.
+def wrap_as_background_process(
+    desc: str,
+) -> Callable[
+    [Callable[P, Awaitable[Optional[R]]]],
+    Callable[P, "defer.Deferred[Optional[R]]"],
+]:
+    """Decorator that wraps an asynchronous function `func`, returning a synchronous
+    decorated function. Calling the decorated version runs `func` as a background
+    process, forwarding all arguments verbatim.
 
-    Equivalent to calling the function with `run_as_background_process`
-    """
+    That is,
+
+        @wrap_as_background_process
+        def func(*args): ...
+        func(1, 2, third=3)
 
-    def wrap_as_background_process_inner(func: F) -> F:
+    is equivalent to:
+
+        def func(*args): ...
+        run_as_background_process(func, 1, 2, third=3)
+
+    The former can be convenient if `func` needs to be run as a background process in
+    multiple places.
+    """
+    # Note: the decorated version of `func` (`wrap_as_background_process_inner`) could be
+    # considered an asynchronous function. However, recent mypy versions warn us that
+    # we have forgotten to `await` a coroutine if we mark the decorated function as
+    # returning an Awaitable.
+    #
+    #     error: Value of type "Coroutine[Any, Any, None]" must be used  [unused-coroutine]
+    #     note: Are you missing an await?
+    # It seems happier when the decorated version of `func` returns a Deferred.
```

Comment:
I think(?) this comes from a mistake in the annotation of `wrap_as_background_process`.

Let's walk through an example. Running mypy on b5228cab0a839110d43025039cfbadc771d39722 contains 10 errors of this form:

```
synapse/app/phone_stats_home.py:181: error: Value of type "Coroutine[Any, Any, None]" must be used  [unused-coroutine]
synapse/app/phone_stats_home.py:181: note: Are you missing an await?
```

The line in question is

https://github.com/matrix-org/synapse/blob/e24ff8ebe3d4119d377355402245947f7de61c00/synapse/app/phone_stats_home.py#L181

`reveal_type(hs.get_datastores().main.reap_monthly_active_users)` yields
```
Revealed type is "def () -> typing.Coroutine[Any, Any, None]"
```
That function is decorated by  `wrap_as_background_process("reap_monthly_active_users")`.

As of this commit, the decorator looks like

https://github.com/matrix-org/synapse/blob/ae01a7edd32c8a4650700294c50a37385fa07984/synapse/metrics/background_process_metrics.py#L259-L278

Here, `F` is bound to the type of the _undecorated_ function

https://github.com/matrix-org/synapse/blob/646324437543c096e737777c81b4fe4b45c3e1a7/synapse/storage/databases/main/monthly_active_users.py#L165-L166

which is a coroutine.

The _decorated_ functoion is `wrap_as_background_process_inner_2` which has been cast to `F`. I think this cast is incorrect(?).

Reply:
> The _decorated_ functoion is `wrap_as_background_process_inner_2` which has been cast to `F`. I think this cast is incorrect(?).

That sounds right. I believe the previous set of type hints were inaccurate because we didn't have `ParamSpec` available at the time?

Reply:
What I think I had in mind:
- the decorated function `wrap_as_background_process_inner_2` returns a `Deferred[Optional[R]]`.
- That type belongs to the more general type of `Awaitable[Optional[R]]`
- I think I might have tried using that more general type as the return type here and still seen those errors.
- But having tried it again just now, I don't get any errors.

I think the comment is confused and should go.